cmd/kthena-router/app/router.go

```diff
@@ -109,15 +155,434 @@ func (s *Server) startRouter(ctx context.Context, router *router.Router, store d
 		}
 	}()
 
-	<-ctx.Done()
-	// graceful shutdown
-	klog.Info("Shutting down HTTP server ...")
-	ctx, cancel := context.WithTimeout(context.Background(), gracefulShutdownTimeout)
-	defer cancel()
-	if err := server.Shutdown(ctx); err != nil {
-		klog.Errorf("Server shutdown failed: %v", err)
+	go func() {
+		<-ctx.Done()
+		// graceful shutdown
+		klog.Info("Shutting down default HTTP server ...")
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), gracefulShutdownTimeout)
+		defer cancel()
+		if err := server.Shutdown(shutdownCtx); err != nil {
+			klog.Errorf("Default server shutdown failed: %v", err)
+		}
+		klog.Info("Default HTTP server exited")
+	}()
+}
+
+// ListenerConfig represents a single listener configuration
+type ListenerConfig struct {
+	GatewayKey   string
+	ListenerName string
+	Port         int32
+	Hostname     *string // nil means match all hostnames
+	Protocol     string
+}
+
+// PortListenerInfo contains all listeners for a specific port
+type PortListenerInfo struct {
+	Server       *http.Server
+	ShutdownFunc context.CancelFunc
+	Listeners    []ListenerConfig
+}
+
+// ListenerManager manages Gateway listeners dynamically
+// Uses port as the key, allowing multiple listeners per port
+type ListenerManager struct {
+	ctx              context.Context
+	router           *router.Router
+	store            datastore.Store
+	server           *Server
+	mu               sync.RWMutex
+	portListeners    map[int32]*PortListenerInfo // key: port
+	gatewayListeners map[string][]ListenerConfig // key: gatewayKey, tracks listeners per gateway
+}
+
+// NewListenerManager creates a new listener manager
+func NewListenerManager(ctx context.Context, router *router.Router, store datastore.Store, server *Server) *ListenerManager {
+	return &ListenerManager{
+		ctx:              ctx,
+		router:           router,
+		store:            store,
+		server:           server,
+		portListeners:    make(map[int32]*PortListenerInfo),
+		gatewayListeners: make(map[string][]ListenerConfig),
+	}
+}
+
+// findBestMatchingListener finds the best matching listener for a request
+// Returns the listener config and true if found, nil and false otherwise
+func (lm *ListenerManager) findBestMatchingListener(port int32, hostname string) (*ListenerConfig, bool) {
+	lm.mu.RLock()
+	defer lm.mu.RUnlock()
+
+	portInfo, exists := lm.portListeners[port]
+	if !exists || len(portInfo.Listeners) == 0 {
+		return nil, false
 	}
-	klog.Info("HTTP server exited")
+
+	// First, try to find an exact hostname match
+	for i := range portInfo.Listeners {
+		listener := &portInfo.Listeners[i]
+		if listener.Hostname != nil && *listener.Hostname == hostname {
+			return listener, true
+		}
+	}
+
+	// If no exact match, try to find a listener without hostname restriction (wildcard)
```

Comment:
Add a todo: wildcard matching is not supported yet

Reply:
fixed.